src/rspq/rspq.c

```diff
@@ -0,0 +1,1235 @@
+    uint32_t slot_count = (command_count + 15) / 16;
+
+    // Reset the overlay descriptor
+    memset(overlay, 0, sizeof(rspq_overlay_t));
+
+    // Remove all registered ids
+    for (uint32_t i = unshifted_id; i < slot_count; i++)
+    {
+        rspq_data.tables.overlay_table[i] = 0;
+    }
+
+    // Reset the command base in the overlay header
+    overlay_header->command_base = 0;
+    data_cache_hit_writeback_invalidate(overlay_header, sizeof(rspq_overlay_header_t));
+
+    rspq_update_tables(false);
+}
+
+/**
+ * @brief Switch to the next write buffer for the current RSP queue.
+ * 
+ * This function is invoked by #rspq_write when the current buffer is
+ * full, that is, when the write pointer (#rspq_cur_pointer) reaches
+ * the sentinel (#rspq_cur_sentinel). This means that we cannot safely
+ * write any more new command in the buffer (the remaining bytes are less
+ * than the maximum command size), and thus a new buffer must be configured.
+ * 
+ * If we're creating a block, we need to allocate a new buffer from the heap.
+ * Otherwise, if we're writing into either the lowpri or the highpri queue,
+ * we need to switch buffer (double buffering strategy), making sure the
+ * other buffer has been already fully executed by the RSP.
+ */
+__attribute__((noinline))
+void rspq_next_buffer(void) {
+    // If we're creating a block
+    if (rspq_block) {
+        // Allocate next chunk (double the size of the current one).
+        // We use doubling here to reduce overheads for large blocks
+        // and at the same time start small.
+        if (rspq_block_size < RSPQ_BLOCK_MAX_SIZE) rspq_block_size *= 2;
+
+        // Allocate a new chunk of the block and switch to it.
+        uint32_t *rspq2 = malloc_uncached(rspq_block_size*sizeof(uint32_t));
+        volatile uint32_t *prev = rspq_switch_buffer(rspq2, rspq_block_size, true);
+
+        // Terminate the previous chunk with a JUMP op to the new chunk.
+        rspq_append1(prev, RSPQ_CMD_JUMP, PhysicalAddr(rspq2));
+        return;
+    }
+
+    // Wait until the previous buffer is executed by the RSP.
+    // We cannot write to it if it's still being executed.
+    // FIXME: this should probably transition to a sync-point,
+    // so that the kernel can switch away while waiting. Even
+    // if the overhead of an interrupt is obviously higher.
+    MEMORY_BARRIER();
+    if (!(*SP_STATUS & rspq_ctx->sp_status_bufdone)) {
+        rspq_flush_internal();
+        RSP_WAIT_LOOP(200) {
+            if (*SP_STATUS & rspq_ctx->sp_status_bufdone)
+                break;
+        }
+    }
+    MEMORY_BARRIER();
+    *SP_STATUS = rspq_ctx->sp_wstatus_clear_bufdone;
+    MEMORY_BARRIER();
+
+    // Switch current buffer
+    rspq_ctx->buf_idx = 1-rspq_ctx->buf_idx;
+    uint32_t *new = rspq_ctx->buffers[rspq_ctx->buf_idx];
+    volatile uint32_t *prev = rspq_switch_buffer(new, rspq_ctx->buf_size, true);
+
+    // Terminate the previous buffer with an op to set SIG_BUFDONE
```

Comment:
Couldn't we append these commands before waiting for the prev buffer to finish?

Reply:
I think @rasky will have to explain this one :)

Reply:
We could append `WRITE_STATUS`, but not `JUMP`, because `JUMP` requires the new buffer to be cleared (or the RSP could pick up the jump before you clear it, and you would end up with the RSP re-execing the previous contents of the new buffer).

Now that I think of it, one option would be to clear just the first word of the new buffer, which is enough to keep the RSP waiting there even in the worst case. So the sequence might be:

 * Clear first word of the new buffer (crossing finger it's been already executed -- very likely but we have no explicit knowledge that this was done)
 * Write the epilogue of the current buffer
 * Wait for the signal to be set
 * Clear the new buffer and set it as current

I'm a little hesitant on this because it'd introduce another delicate synchronization process, for something which I feel might be a marginal performance improvement. I would say the largest time that you would parallelize to RSP is the memset that clears the new buffer, which I previously already attempted to parallelize by using DMA, but then I decided to punt for simplicity.